common/desktop-exports

```diff
@@ -102,20 +104,62 @@ mkdir -p $XDG_CONFIG_HOME
 # Ensure the app finds locale definitions (requires locales-all to be installed)
 append_dir LOCPATH $RUNTIME/usr/lib/locale
 
-# Run xdg-user-dirs-update
-if [ `which xdg-user-dirs-update` ]; then
-   xdg-user-dirs-update
+# If any, keep track of where XDG dirs were so we can potentially migrate the content later
+test -f ${XDG_CONFIG_HOME:-$HOME/.config}/user-dirs.dirs && . ${XDG_CONFIG_HOME:-$HOME/.config}/user-dirs.dirs
+for d in DOCUMENTS DESKTOP DOWNLOAD MUSIC PICTURES VIDEOS PUBLICSHARE TEMPLATES; do
+  eval $(echo "OLD_XDG_${d}_DIR")=`eval "$(echo "echo \\$XDG_${d}_DIR")"`
+done
+
+# Setup user-dirs.* or run xdg-user-dirs-update if needed
+needs_xdg_update=false
+needs_xdg_links=false
+if can_open_file "$REALHOME/.config/user-dirs.dirs" && can_open_file "$REALHOME/.config/user-dirs.locale"; then
+   sed s#\$HOME#${REALHOME}#g $REALHOME/.config/user-dirs.dirs > $HOME/.config/user-dirs.dirs
+   cp -a $REALHOME/.config/user-dirs.locale $HOME/.config/
+   for f in user-dirs.dirs user-dirs.locale; do
+     md5sum < $REALHOME/.config/$f > $HOME/.config/$f.md5sum
+   done
+else;
+   needs_xdg_update=true
+   needs_xdg_links=true
 fi
 
-# Create links for user-dirs.dirs
+# Check if we can actually read the contents of each xdg dir
 test -f ${XDG_CONFIG_HOME:-$HOME/.config}/user-dirs.dirs && . ${XDG_CONFIG_HOME:-$HOME/.config}/user-dirs.dirs
 XDG_SPECIAL_DIRS=($XDG_DOCUMENTS_DIR $XDG_DESKTOP_DIR $XDG_DOWNLOAD_DIR $XDG_MUSIC_DIR $XDG_PICTURES_DIR $XDG_VIDEOS_DIR $XDG_PUBLIC_DIR $XDG_TEMPLATES_DIR)
 for d in ${XDG_SPECIAL_DIRS[@]}; do
-  b=$(basename "$d")
+  if ! can_open_file $d; then
+    needs_xdg_update=true
+  fi
+done
+
+# If needs XDG update and xdg-user-dirs-update exists in $PATH, run it
+if [ $needs_xdg_update = true && `which xdg-user-dirs-update` ]; then
+   xdg-user-dirs-update
+fi
+    
+# Create links for user-dirs.dirs
+if [ $needs_xdg_links = true ]; then
+  test -f ${XDG_CONFIG_HOME:-$HOME/.config}/user-dirs.dirs && . ${XDG_CONFIG_HOME:-$HOME/.config}/user-dirs.dirs
+  XDG_SPECIAL_DIRS=($XDG_DOCUMENTS_DIR $XDG_DESKTOP_DIR $XDG_DOWNLOAD_DIR $XDG_MUSIC_DIR $XDG_PICTURES_DIR $XDG_VIDEOS_DIR $XDG_PUBLIC_DIR $XDG_TEMPLATES_DIR)
+  for d in ${XDG_SPECIAL_DIRS[@]}; do
+    b=$(basename "$d")
     if [ -e $REALHOME/$b ] && [ ! -e $HOME/$b ]; then
       ln -s $REALHOME/$b $HOME/$b
     fi
-done
+  done
+else
+  # If we aren't creating new links, check if we have content saved in old locations and move it
+  for d in DOCUMENTS DESKTOP DOWNLOAD MUSIC PICTURES VIDEOS PUBLICSHARE TEMPLATES; do
+    old=`eval "$(echo "echo \\$OLD_XDG_${d}_DIR")"`
+    new=`eval "$(echo "echo \\$XDG_${d}_DIR")"`
+    if [ -L "$old" ] && [ -d "$new" ] && [ `readlink "$old"` != "$new" ]; then
+      mv "$old"/* "$new"/
+    elif [ -d "$old" ] && [ -d "$new" ] && [ "$old" != "$new" ]; then
+      mv "$old"/* "$new"/
+    fi
```

Comment:
I rebuilt the chromium snap with those changes, and the first time I launched it I got a number of message triggered by those calls to mv:

mv: '/home/osomon/snap/chromium/x2/Documents/some-file.pdf' and '/home/osomon/Documents/some-file.pdf' are the same file
(repeat the above for all the files under $HOME/Documents)

and then:

mv: cannot stat '/home/osomon/snap/chromium/x2/Desktop/*': No such file or directory
mv: cannot stat '/home/osomon/snap/chromium/x2/Music/*': No such file or directory
mv: cannot stat '/home/osomon/snap/chromium/x2/Pictures/*': No such file or directory
mv: cannot stat '/home/osomon/snap/chromium/x2/Templates/*': No such file or directory

This didn't seem to break anything, but I think we should try and avoid those messages if possible, otherwise users might think something is broken.

Reply:
Fixed